File: test/optimization/NonlinearProblemTest.cpp
// Copyright (c) Joshua Nichols and Tyler Veness

#include <vector>

#include "gtest/gtest.h"
#include "sleipnir/optimization/OptimizationProblem.h"

std::vector<double> Range(double start, double end, double step) {
  std::vector<double> ret;

  for (double i = start; i < end; i += step) {
    ret.emplace_back(i);
  }

  return ret;
}

TEST(NonlinearProblemTest, Quartic) {
  sleipnir::OptimizationProblem problem;

  auto x = problem.DecisionVariable();
  x = 20.0;

  problem.Minimize(sleipnir::pow(x, 4));

  problem.SubjectTo(x >= 1);

  sleipnir::SolverConfig config;
  config.diagnostics = true;

  auto status = problem.Solve(config);

  EXPECT_EQ(sleipnir::autodiff::ExpressionType::kNonlinear,
            status.costFunctionType);
  EXPECT_EQ(sleipnir::autodiff::ExpressionType::kNone,
            status.equalityConstraintType);
  EXPECT_EQ(sleipnir::autodiff::ExpressionType::kLinear,
            status.inequalityConstraintType);
  EXPECT_EQ(sleipnir::SolverExitCondition::kOk, status.exitCondition);

  EXPECT_NEAR(1.0, x.Value(0), 1e-6);
}

TEST(NonlinearProblemTest, DISABLED_RosenbrockWithCubicAndLineConstraint) {
  // https://en.wikipedia.org/wiki/Test_functions_for_optimization#Test_functions_for_constrained_optimization
  for (auto x0 : Range(-1.5, 1.5, 0.1)) {
    for (auto y0 : Range(-0.5, 2.5, 0.1)) {
      sleipnir::OptimizationProblem problem;

      auto x = problem.DecisionVariable();
      x = x0;
      auto y = problem.DecisionVariable();
      y = y0;

      problem.Minimize(sleipnir::pow(1 - x, 2) +
                       100 * sleipnir::pow(y - sleipnir::pow(x, 2), 2));

      problem.SubjectTo(sleipnir::pow(x - 1, 3) - y + 1 <= 0);
      problem.SubjectTo(x + y - 2 <= 0);

      auto status = problem.Solve();

      EXPECT_EQ(sleipnir::autodiff::ExpressionType::kNonlinear,
                status.costFunctionType);
      EXPECT_EQ(sleipnir::autodiff::ExpressionType::kNone,
                status.equalityConstraintType);
      EXPECT_EQ(sleipnir::autodiff::ExpressionType::kNonlinear,
                status.inequalityConstraintType);
      EXPECT_EQ(sleipnir::SolverExitCondition::kOk, status.exitCondition);

      auto Near = [](double expected, double actual, double tolerance) {
        return std::abs(expected - actual) < tolerance;
      };

      // Local minimum at (0.0, 0.0)
      // Global minimum at (1.0, 1.0)
      EXPECT_TRUE(Near(0.0, x.Value(0), 1e-2) || Near(1.0, x.Value(0), 1e-2))
          << fmt::format("  (x₀, y₀) = ({}, {})\n", x0, y0)
          << fmt::format("  x.Value(0) = {}", x.Value(0));
      EXPECT_TRUE(Near(0.0, y.Value(0), 1e-2) || Near(1.0, y.Value(0), 1e-2))
          << fmt::format("  (x₀, y₀) = ({}, {})\n", x0, y0)
          << fmt::format("  y.Value(0) = {}", y.Value(0));
    }
  }
}

TEST(NonlinearProblemTest, RosenbrockWithDiskConstraint) {
  // https://en.wikipedia.org/wiki/Test_functions_for_optimization#Test_functions_for_constrained_optimization
  for (auto x0 : Range(-1.5, 1.5, 0.1)) {
    for (auto y0 : Range(-1.5, 1.5, 0.1)) {
      sleipnir::OptimizationProblem problem;

      auto x = problem.DecisionVariable();
      x = x0;
      auto y = problem.DecisionVariable();
      y = y0;

      problem.Minimize(sleipnir::pow(1 - x, 2) +
                       100 * sleipnir::pow(y - sleipnir::pow(x, 2), 2));

      problem.SubjectTo(sleipnir::pow(x, 2) + sleipnir::pow(y, 2) <= 2);

      auto status = problem.Solve();

      EXPECT_EQ(sleipnir::autodiff::ExpressionType::kNonlinear,
                status.costFunctionType);
      EXPECT_EQ(sleipnir::autodiff::ExpressionType::kNone,
                status.equalityConstraintType);
      EXPECT_EQ(sleipnir::autodiff::ExpressionType::kQuadratic,
                status.inequalityConstraintType);
      EXPECT_EQ(sleipnir::SolverExitCondition::kOk, status.exitCondition);

      EXPECT_NEAR(1.0, x.Value(0), 1e-1)
          << fmt::format("  (x₀, y₀) = ({}, {})\n", x0, y0)
          << fmt::format("  x.Value(0) = {}", x.Value(0));
      EXPECT_NEAR(1.0, y.Value(0), 1e-1)
          << fmt::format("  (x₀, y₀) = ({}, {})\n", x0, y0)
          << fmt::format("  x.Value(0) = {}", x.Value(0));
    }
  }
}

File: jormungandr/test/optimization/nonlinear_problem_test.py
import platform

import numpy as np
import pytest

import jormungandr.autodiff as autodiff
from jormungandr.autodiff import ExpressionType
from jormungandr.optimization import ExitStatus, Problem


def test_quartic():
    problem = Problem()

    x = problem.decision_variable()
    x.set_value(20.0)

    problem.minimize(autodiff.pow(x, 4))

    problem.subject_to(x >= 1)

    assert problem.cost_function_type() == ExpressionType.NONLINEAR
    assert problem.equality_constraint_type() == ExpressionType.NONE
    assert problem.inequality_constraint_type() == ExpressionType.LINEAR

    assert problem.solve(diagnostics=True) == ExitStatus.SUCCESS

    assert x.value() == pytest.approx(1.0, abs=1e-6)


def test_rosenbrock_with_cubic_and_line_constraint():
    # https://en.wikipedia.org/wiki/Test_functions_for_optimization#Test_functions_for_constrained_optimization
    for x0 in np.arange(-1.5, 1.5, 0.1):
        for y0 in np.arange(-0.5, 2.5, 0.1):
            problem = Problem()

            x = problem.decision_variable()
            x.set_value(x0)
            y = problem.decision_variable()
            y.set_value(y0)

            problem.minimize(
                100 * autodiff.pow(y - autodiff.pow(x, 2), 2) + autodiff.pow(1 - x, 2)
            )

            problem.subject_to(y >= autodiff.pow(x - 1, 3) + 1)
            problem.subject_to(y <= -x + 2)

            assert problem.cost_function_type() == ExpressionType.NONLINEAR
            assert problem.equality_constraint_type() == ExpressionType.NONE
            assert problem.inequality_constraint_type() == ExpressionType.NONLINEAR

            assert problem.solve() == ExitStatus.SUCCESS

            # Local minimum at (0.0, 0.0)
            # Global minimum at (1.0, 1.0)
            assert x.value() == pytest.approx(
                0.0, abs=1e-2
            ) or x.value() == pytest.approx(1.0, abs=1e-2)
            assert y.value() == pytest.approx(
                0.0, abs=1e-2
            ) or y.value() == pytest.approx(1.0, abs=1e-2)


def test_rosenbrock_with_disk_constraint():
    # https://en.wikipedia.org/wiki/Test_functions_for_optimization#Test_functions_for_constrained_optimization
    for x0 in np.arange(-1.5, 1.5, 0.1):
        for y0 in np.arange(-1.5, 1.5, 0.1):
            problem = Problem()

            x = problem.decision_variable()
            x.set_value(x0)
            y = problem.decision_variable()
            y.set_value(y0)

            problem.minimize(
                100 * autodiff.pow(y - autodiff.pow(x, 2), 2) + autodiff.pow(1 - x, 2)
            )

            problem.subject_to(autodiff.pow(x, 2) + autodiff.pow(y, 2) <= 2)

            assert problem.cost_function_type() == ExpressionType.NONLINEAR
            assert problem.equality_constraint_type() == ExpressionType.NONE
            assert problem.inequality_constraint_type() == ExpressionType.QUADRATIC

            assert problem.solve() == ExitStatus.SUCCESS

            assert x.value() == pytest.approx(1.0, abs=1e-3)
            assert y.value() == pytest.approx(1.0, abs=1e-3)


def test_minimum_2d_distance_with_linear_constraint():
    problem = Problem()

    x = problem.decision_variable()
    x.set_value(20.0)

    y = problem.decision_variable()
    y.set_value(50.0)

    problem.minimize(autodiff.sqrt(x * x + y * y))

    problem.subject_to(y == -x + 5.0)

    assert problem.cost_function_type() == ExpressionType.NONLINEAR
    assert problem.equality_constraint_type() == ExpressionType.LINEAR
    assert problem.inequality_constraint_type() == ExpressionType.NONE

    if platform.system() == "Linux" and platform.machine() == "aarch64":
        # FIXME: Fails on Linux aarch64 with "line search failed"
        assert problem.solve(diagnostics=True) == ExitStatus.LINE_SEARCH_FAILED
        return
    else:
        assert problem.solve(diagnostics=True) == ExitStatus.SUCCESS

    assert x.value() == pytest.approx(2.5, abs=1e-2)
    assert y.value() == pytest.approx(2.5, abs=1e-2)


def test_wachter_and_biegler_line_search_failure():
    # See example 19.2 of [1]

    problem = Problem()

    x = problem.decision_variable()
    s1 = problem.decision_variable()
    s2 = problem.decision_variable()

    x.set_value(-2)
    s1.set_value(3)
    s2.set_value(1)

    problem.minimize(x)

    problem.subject_to(x**2 - s1 - 1 == 0)
    problem.subject_to(x - s2 - 0.5 == 0)
    problem.subject_to(s1 >= 0)
    problem.subject_to(s2 >= 0)

    assert problem.cost_function_type() == ExpressionType.LINEAR
    assert problem.equality_constraint_type() == ExpressionType.QUADRATIC
    assert problem.inequality_constraint_type() == ExpressionType.LINEAR

    # FIXME: Fails with "line search failed"
    assert problem.solve(diagnostics=True) == ExitStatus.LINE_SEARCH_FAILED
    return

    assert x.value() == 1.0
    assert s1.value() == 0.0
    assert s2.value() == 0.5
